### PR TITLE
Greatly optimize filter and match performance

### PR DESCRIPTION
**Pull request by M09Ic**
example data:
ip.txt:
```
1.25.242.23/32
1.71.150.96/29
54.240.128.0/20
```

cdn:

```
109.199.57.0/24
101.79.149.0/24
116.128.128.0/24
115.231.186.0/24
103.22.200.0/22
61.155.165.0/24
113.20.148.0/23
185.254.242.0/24
59.36.120.0/24
70.132.0.0/18
52.78.247.128/26
110.93.176.0/20
54.240.128.0/18
46.51.216.0/21
70.132.0.0/16
104.16.0.0/12
121.12.98.0/24
103.31.4.0/22
204.246.164.0/22
42.236.6.0/24
125.88.190.0/24
61.54.47.0/24
103.27.12.0/22
116.193.80.0/21
131.0.72.0/22
192.230.64.0/18
220.191.231.0/24
59.82.0.0/16
```

old:
```
$ time mapcidr -cl ip.txt  -fi cdn.txt  -silent
1.25.242.23
1.71.150.96
1.71.150.97
1.71.150.98
1.71.150.99
1.71.150.100
1.71.150.101
1.71.150.102
1.71.150.103

real    0m11.365s
user    0m0.000s
sys     0m0.000s
```

optimizated:
```
$ time mapcidr -cl ip.txt  -fi cdn.txt  -silent
1.25.242.23/32
1.71.150.96/29

real    0m0.051s
user    0m0.000s
sys     0m0.016s
```

Time complexity from n^n to n*n (mybe, I did not calculate in detail)

### CODE REVIEW

File: cidr.go [hunk 1]
```diff
@@ -127,6 +127,22 @@ func SplitIPNetIntoN(iprange *net.IPNet, n int) ([]*net.IPNet, error) {
...
+// DifferenceCIDR set difference of target and exclude
+func DifferenceCIDR(target, exclude *net.IPNet) []*net.IPNet {
+	left, _, right := partitionCIDR(*target, *exclude)
+	return append(left, right...)
+}
+
+// ContainsCIDR check if A contains B
+func ContainsCIDR(A, B *net.IPNet) bool {
+	Bfirst, Blast, _ := AddressRange(B)
+	if A.Contains(Bfirst) && A.Contains(Blast) {
```
Comment (Mzack9999) [inline cidr.go:139]: I think we can use:
```go
return A.Contains(Bfirst) && A.Contains(Blast)
```

File: cmd/mapcidr/main.go [hunk 2]
```diff
@@ -107,8 +107,8 @@ func ParseOptions() *Options {
 
 	// Miscellaneous
 	flagSet.CreateGroup("miscellaneous", "Miscellaneous",
-		flagSet.BoolVarP(&options.SortAscending, "sort", "s", false, "Sort input IPs/CIDRs in ascending order"),
-		flagSet.BoolVarP(&options.SortDescending, "sort-reverse", "sr", false, "Sort input IPs/CIDRs in descending order"),
+		flagSet.BoolVarP(&options.SortAscending, "sort", "s", false, "Sort input IPs in ascending order"),
```
Comment (Mzack9999) [inline cmd/mapcidr/main.go:110]: it seems to me that both CIDRs and IPs are supported?

**Review by Mzack9999 — CHANGES_REQUESTED** [review]

File: cidr.go (hunk 1 above)
Comment (M09Ic) [reply cidr.go:139]: No, Consider two examples.

### 1. 
If ip is less than cdn, like.

ip.txt: `1.1.1.1/28`

cdn.txt: `1.1.1.1/24`

no output, `return A.Contains(Bfirst) && A.Contains(Blast)`  is correct,

### 2.
but If If ip is greater than cdn.

ip.txt: `1.1.1.1/20`

cdn.txt: `1.1.1.1/24`

We need to subtract /24 from /20.
output:
```
1.1.0.0/24
1.1.8.0/21
1.1.4.0/22
1.1.2.0/23
```

File: cmd/mapcidr/main.go (hunk 2 above)
Comment (M09Ic) [reply cmd/mapcidr/main.go]: Sorry, I'm not very good with git, I've been using my previous code to make changes, which caused some of the code to be overwritten with past versions.

Now, I've reverted the changes and committed

Comment (M09Ic) [reply cmd/mapcidr/main.go]: This commit is only about match and filter.

File: cidr.go (hunk 1 above)
Comment (Mzack9999) [reply cidr.go:139]: If I'm not mistaken, the output of the following code snippet:
```go
if A.Contains(Bfirst) && A.Contains(Blast) { // true if both conditions are matched
	return true // both conditions matched
} else {
	return false
}
```
is equivalent to directly return the result of the boolean comparison:
```go
return A.Contains(Bfirst) && A.Contains(Blast)
```

Comment (M09Ic) [reply cidr.go]: You are right. I misunderstood the code snippet

**Review by Mzack9999 — COMMENTED** [review]
This PR will need extensive testing. Some preliminary observations:
- Some integration tests are failing
- it seems like IPv4-mapped IPv6 addresses are now represented only as ipv4 (see failing integration tests)
- Results tends to be cidr

**Comment by M09Ic** [comment]
https://github.com/projectdiscovery/mapcidr/pull/107/commits/53d4cf19f25c254cdee892156866c60db21d614a Resolved. 

test case1:
> Some integration tests are failing

and 

> Results tends to be cidr

`mapcidr -cl ip.txt -fi 192.168.1.1,192.168.1.3 -silent`

ip.txt:
```
192.168.1.0/30
```
output:
```
192.168.1.0
192.168.1.2
```

test case2:

> it seems like IPv4-mapped IPv6 addresses are now represented only as ipv4 (see failing integration tests)

`mapcidr -cl ip.txt -fi cdn.txt -silent`

ip.txt:
```
2002:0000:0000:0000:0000:0000:c0a8:0101
2002:0000:0000:0000:0000:0000:c0a8:0104
```
cdn.txt:
```
2002:0000:0000:0000:0000:0000:c0a8:0101
2002:0000:0000:0000:0000:0000:c0a8:0103
```
output:
```
2002::c0a8:104
```

**Review by Mzack9999 — CHANGES_REQUESTED** [review]
Additional integration tests failing, apparently the change implemented impact a lot of functionalities. I'll try to deeply review it in upcoming days, marking with "request changes"

**Comment by M09Ic** [comment]
now, all integration tests pass, lol

**Review by M09Ic — COMMENTED** [review]
done

**Comment by Mzack9999** [comment]
While testing ipv4/ipv6, further inconsistencies were detected - For the time being. I'm closing it as not planned for the time being as the library it's heavily used with other tools.

**Pull request closed without merging**